### ISSUE TITLE
torch 2.14 flags a singular matrix that 2.9.1 did not, and symmetric_transfer_error returns NaN instead of max_num for a singular H

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4203]
### Summary

Two linked findings, surfaced while moving the MPS gate to torch 2.14.0 (#4202).

1. **torch 2.14.0 detects an exactly-singular float32 matrix that 2.9.1 did not.** kornia supports
   `torch>=2.5.1`, so the two ends of the supported range disagree about whether the same input is
   invertible, and kornia branches on that answer.
2. **`symmetric_transfer_error` returns `nan` instead of `max_num` for a singular homography.**
   The mask arithmetic cannot neutralise a non-finite inverse, which is precisely the case the
   mask exists for. This one is not version-specific; finding 1 just makes it reachable for more
   inputs.

### 1. The cross-version difference

The matrix (rows 0 and 2 identical, so `det` is exactly 0):

```python
A = torch.tensor([[-956.0, -761.0, -67.0], [760.0, -37.0, -621.0], [-956.0, -761.0, -67.0]])
```

| | torch 2.9.1 | torch 2.14.0 |
|---|---|---|
| `torch.linalg.inv_ex(A).info` | `0` (reported valid) | **`3`** |
| `torch.linalg.inv(A)` | returned, no exception | **raises `_LinAlgError`** |
| `torch.linalg.det(A)` | `1.969` | `1.969` |

Same machine (M1, macOS 26.5.1), same fp32 input, measured a few hours apart on 2.9.1 and 2.14.0.
`det` still misreports on both — a zero determinant was never a reliable singularity test — but
the LU-based detection changed.

Consequences inside kornia:

- `kornia.core.utils._torch_inverse_cast` now **raises** `_LinAlgError` for this input where it
  previously returned a non-finite matrix. Everything that goes through it inherits that:
  `normalize_homography` / `denormalize_homography`, `invert_affine_transform`,
  `PinholeCamera.intrinsics_inverse`, every geometric augmentation's inverse transform,
  `image_registrator`, `planar_tracker`.
- `kornia.core.utils.safe_inverse_with_mask` returns `mask=False` where it returned `mask=True`,
  so `symmetric_transfer_error`, and therefore `find_homography_dlt_iterated` and `RANSAC`
  homography scoring, classify hypotheses differently between supported torch versions.
- The export path disagrees with the eager path for the same input: `safe_inverse_with_mask`
  under tracing takes the closed-form branch, whose validity test is `det != 0`, and the
  closed-form determinant of this matrix is `-20.0`, so tracing says **valid** while eager says
  **invalid**.

Verified on 2.14.0:

```
linalg.det                    = 1.969299554824829
inv_ex info                   = 3
_torch_inverse_cast           -> raised _LinAlgError
safe_inverse_with_mask mask   = [False]
closed-form det (export path) = [-20.0]  -> mask [True]
```

### 2. The masking bug this exposes

`symmetric_transfer_error` documents that a non-invertible `H` scores `max_num`:

```python
H_inv, good_H = safe_inverse_with_mask(H)
there = oneway_transfer_error(pts1, pts2, H, True, eps)
back  = oneway_transfer_error(pts2, pts1, H_inv, True, eps)
good_H_reshape = good_H.view(-1, 1).expand_as(there)
out = (there + back) * good_H_reshape.to(there.dtype) + max_num * (~good_H_reshape).to(there.dtype)
```

When `good_H` is `False`, `H_inv` is exactly what `inv_ex` leaves behind for a singular input —
here `[nan, nan, nan, inf, ...]`. `back` is then non-finite, and `nan * 0.0` is `nan`, not `0.0`,
so `out` is `nan` rather than `max_num`. Reproduced on 2.14.0 with two different singular
homographies, including a trivially singular one that every torch version flags:

```python
p1, p2 = torch.rand(1, 5, 2), torch.rand(1, 5, 2)
Z = torch.zeros(1, 3, 3); Z[0, 0, 0] = 1.0
symmetric_transfer_error(p1, p2, Z)   # -> [nan, nan, nan, nan, nan], expected 3.4028235e+38
```

`find_homography_dlt_iterated` feeds that straight into
`weights_new = torch.exp(-errors / (2 * soft_inl_th**2))`, so one singular intermediate `H` turns
every weight into `nan` and the returned homography with it — silently, since nothing raises.

The fix is `torch.where(good_H_reshape, there + back, max_num)`, which never multiplies a
non-finite value by zero. Worth checking the other error functions in `kornia/geometry/ransac.py`
for the same pattern.

### Suggested scope

- Replace the arithmetic mask in `symmetric_transfer_error` with `torch.where`, with a test that
  pins `max_num` for a singular `H` (the current shape-only coverage passes on `nan`).
- Decide the contract for `_torch_inverse_cast` on a singular input now that 2.5.1 and 2.14.0
  disagree: raise on both, or mask on both. Whatever is chosen, it should not depend on the torch
  version the user installed.
- Reconcile the eager and traced branches of `safe_inverse_with_mask`, which currently return
  opposite validity for the same matrix.

### Environment

kornia `main`, torch 2.14.0 and 2.9.1, macOS 26.5.1 arm64 (Apple M1), Python 3.11, CPU float32.
Related: #4201 (MPS SVD ceiling), #4202 (MPS gate on 2.14.0).

Posted on behalf of @ducha-aiki by Claude (Claude Opus 5).

**Comment by cris-pat-37** [comment]
Hi @ducha-aiki , I’d like to work on the symmetric_transfer_error masking bug described in #4203. I’ve reproduced the NaN result for singular homographies and understand the proposed torch.where fix. I’ll keep the scope focused on symmetric_transfer_error and add a regression test ensuring singular homographies return max_num. Would you be okay with me taking this part?
